### PR TITLE
Api 1640 token searching

### PR DESCRIPTION
**Pull request by jhulbert-va**
https://vajira.max.gov/browse/API-1640

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportController.java [hunk 1]
```diff
@@ -93,13 +98,61 @@
     return bundle(parameters, fhir, (int) entitiesPage.getTotalElements());
   }
 
+  private Set<DiagnosticReportEntity.CategoryCodes> categoriesFor(TokenParameter categoryToken) {
+    return categoryToken
+        .behavior()
+        .onExplicitSystemAndExplicitCode((s, c) -> datamartCategoryCodesFor(c))
+        .onAnySystemAndExplicitCode(this::datamartCategoryCodesFor)
+        .onNoSystemAndExplicitCode(this::datamartCategoryCodesFor)
```
Comment (eclendenning-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportController.java:106]: Explicitly no system should be an empty bundle.

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java [hunk 2]
```diff
@@ -567,40 +608,55 @@ void searchByPatientAndEmptyCodeAndNoDate() {
...
-                        "http://fonzy.com/cool/DiagnosticReport" + "?category=NOPE" + "&patient=p0",
-                        0,
+                        "http://fonzy.com/cool/DiagnosticReport?patient=p0&status=final",
+                        1,
                         15))));
   }
 
   @Test
-  void searchByPatientAndStatus() {
+  void searchByPatientAndStatusWithBadSystemAndAnyCode() {
```
Comment (bschofield-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java:642]: This is not a _bad_ system, it's an _unknown_ system and should result in an empty bundle

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java [hunk 3]
```diff
@@ -419,6 +419,47 @@ void searchByPatientAndCategoryAndTwoDates() {
...
 
+  @Test
+  void searchByPatientAndCategoryWithBadSystemAndAnyCode() {
+    assertThrows(
+        IllegalStateException.class,
+        () -> controller().searchByPatientAndCategory("x", "http://not.today|", null, 1, 15));
+  }
+
+  @Test
+  void searchByPatientAndCategoryWithExplicitSystem() {
```
Comment (bschofield-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java:430]: This is not a _bad_ system, it's an _unknown_ system and should result in an empty bundle

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java [hunk 4]
```diff
@@ -419,6 +419,47 @@ void searchByPatientAndCategoryAndTwoDates() {
                         15))));
   }
 
+  @Test
+  void searchByPatientAndCategoryWithBadSystemAndAnyCode() {
```
Comment (bschofield-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java:423]: This is not a _bad_ system, it's an _unknown_ system and should result in an empty bundle

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java [hunk 5]
```diff
@@ -419,6 +419,47 @@ void searchByPatientAndCategoryAndTwoDates() {
...
+
+  @Test
+  void searchByPatientAndCategoryWithExplicitSystem() {
+    assertThrows(
+        IllegalStateException.class,
+        () -> controller().searchByPatientAndCategory("x", "http://not.today|", null, 1, 15));
+  }
+
+  @Test
+  void searchByPatientAndCategoryWithInvalidCode() {
```
Comment (bschofield-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java:437]: The code is _unknown_ not _invalid_.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportController.java (hunk 1 above)
Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportController.java:106]: Should be dealt with in `searchByPatientAndCategory()`

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java [hunk 6]
```diff
@@ -420,44 +421,65 @@ void searchByPatientAndCategoryAndTwoDates() {
...
+    assertThat(
+            json(
+                controller()
+                    .searchByPatientAndCategory(
+                        "p0", "http://terminology.hl7.org/CodeSystem/v2-0074|NOPE", null, 1, 15)))
+        .isEqualTo(
+            json(
+                emptyBundleOf(
+                    "category=http://terminology.hl7.org/CodeSystem/v2-0074|NOPE&patient=p0")));
+    // Any System, Unknown Code
```
Comment (eclendenning-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/diagnosticreport/R4DiagnosticReportControllerTest.java:444]: Any system is `code` 
No system is `|code`

**Pull request merged**